### PR TITLE
ZCP 

### PR DESCRIPTION
**Pull request by kshitij-05**
Created a new branch for ZCP, starting out with unittests to make changes to make CP class usable for complex as well as doubles.

### CODE REVIEW

File: btas/tensor.h [hunk 1]
```diff
@@ -22,693 +22,564 @@ namespace btas {
...
+    Tensor(const Tensor& x) : range_(x.range()), storage_(x.storage_) {}
+
+    /// move constructor
+    Tensor(Tensor&& x) : range_(std::move(x.range())), storage_(std::move(x.storage_)) {}
+
+    /// @return deep copy of `*this`, even if `storage_type` is shallow copy
+    Tensor clone() const { return Tensor(range(), storage_type(storage().cbegin(), storage().cend())); }
+
+    /// returns element-wise conjugate of a *this tensor if have complex value type
+    std::conditional_t<is_complex_type_v<_T>, Tensor, const Tensor&> conj() {
```
Comment (evaleev) [inline btas/tensor.h:196]: must be `const` function

Comment (evaleev) [reply btas/tensor.h:196]: review your changes for const correctness

**Comment by evaleev** [comment]
please address issue exposed here: https://github.com/ValeevGroup/BTAS/actions/runs/3875922259/jobs/6609120742#step:10:113

File: btas/tensor.h (hunk 1 above)
Comment (kshitij-05) [reply btas/tensor.h]: added const in latest commit

File: btas/generic/cp_als.h [hunk 2]
```diff
@@ -700,16 +686,15 @@ namespace btas {
       }
       KRP_dims.push_back(ndim);
 
-      contract(1.0, tensor_ref, tref_indices, KhatriRao, KRP_dims, 0.0, temp, An_indices);
+      contract(this-> one , tensor_ref, tref_indices, KhatriRao, KRP_dims, this-> zero, temp, An_indices);
```
Comment (evaleev) [inline btas/generic/cp_als.h:689]: spaces between `->` and member name as well as between `one` and the comma look weird.

File: btas/generic/cp_als.h [hunk 3]
```diff
@@ -818,14 +803,14 @@ namespace btas {
         // over the middle dimension and sum over the rank.
 
         else if (contract_dim > n) {
-          middle_contract(1.0, temp, a, 0.0, contract_tensor);
+          middle_contract(this->one, temp, a.conj(), this-> zero, contract_tensor);
```
Comment (evaleev) [inline btas/generic/cp_als.h:806]: weird spaces ... fix throughout

**Review by evaleev — CHANGES_REQUESTED** [review]

File: btas/generic/cp_als.h (hunk 3 above)
Comment (kshitij-05) [reply btas/generic/cp_als.h]: Done

File: btas/generic/cp_als.h [hunk 4]
```diff
@@ -779,7 +764,7 @@ namespace btas {
                 Range1{last_dim ? size / tensor_ref.extent(contract_dim) : tensor_ref.extent(contract_dim)}});
 
       // contract tensor ref and the first factor matrix
-      gemm((last_dim ? blas::Op::Trans : blas::Op::NoTrans), blas::Op::NoTrans, 1.0, tensor_ref, A[contract_dim], 0.0,
+      gemm((last_dim ? blas::Op::Trans : blas::Op::NoTrans), blas::Op::NoTrans, this-> one , (last_dim? tensor_ref.conj():tensor_ref), A[contract_dim].conj(), this-> zero,
```
Comment (evaleev) [inline btas/generic/cp_als.h:767]: spaces still ...

**Pull request merged**